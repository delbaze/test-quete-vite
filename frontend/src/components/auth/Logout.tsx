import { useAuth } from "@/context/AuthContext";
import { LOGOUT } from "@/requetes/queries/auth.queries";
import { LogoutQuery, LogoutQueryVariables } from "@/types/graphql";
import { useQuery } from "@apollo/client";

function Logout() {
  const {getInfos} = useAuth();
  const { loading } = useQuery<LogoutQuery, LogoutQueryVariables>(LOGOUT, {
    fetchPolicy: "no-cache",
    async onCompleted() {
      await getInfos()
      //libre à vous ensuite de rediriger la personne quelque part... (en utilisant useNavigate)
    },
  });
  return (
    <main
      className={`flex min-h-screen flex-col items-center justify-between p-24`}
    >
      {loading ? "Veuillez patienter..." : "Vous êtes déconnectés!"}
    </main>
  );
}

export default Logout;
