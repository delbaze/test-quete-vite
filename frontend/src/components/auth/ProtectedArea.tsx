import { CHECK_TOKEN } from "@/requetes/queries/auth.queries";
import { CheckTokenQuery } from "@/types/graphql";
import { useQuery } from "@apollo/client";
import { PropsWithChildren } from "react";
import { useLocation, useNavigate } from "react-router-dom";

//pour l'instant je prévoir un énum comme ça, mais lorsque le back aura cette fonctionnalité, on prendra le type depuis codegen
enum Role {
  "ADMIN",
  "USER",
}

function ProtectedArea({
  children,
  role = Role["USER"], // par défaut le rôle est USER
}: PropsWithChildren & { role?: Role }) {
  console.log("%c⧭", "color: #00e600", role);
  const navigate = useNavigate();
  const location = useLocation();
  const { loading } = useQuery<CheckTokenQuery>(CHECK_TOKEN, {
    onError(error) {
      console.log(error); //on verra comment on gère l'erreur
    },
    onCompleted(data) {
      console.log("%c⧭", "color: #00a3cc", data);
      if (!data.checkToken) {
        console.log(location);
        //si la personne n'est pas connectée, on la redirige vers /auth/login, et on envoie désormais la route initialement jointe
        navigate("/auth/login", { state: { initialRoute: location.pathname } });
        return;
      }
      //ici on imaginera stocker dans le contexte les infos contenu dans data

      //on pourra même imaginer regarder le rôle du user
    },
    fetchPolicy: "no-cache",
  });

  if (loading) {
    return <div>Chargement en cours...</div>;
  }

  return <div>{children}</div>;
}

export default ProtectedArea;
