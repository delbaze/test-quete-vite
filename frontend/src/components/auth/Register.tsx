import { REGISTER } from "@/requetes/mutations/auth.mutations";
import {
  InputRegister,
  RegisterMutation,
  RegisterMutationVariables,
} from "@/types/graphql";
import { useMutation } from "@apollo/client";
import { useNavigate } from "react-router-dom";

function Register() {
  const navigate = useNavigate();
  const [register, { error }] = useMutation<
    RegisterMutation,
    RegisterMutationVariables
  >(REGISTER, {
    onCompleted: (data) => {
      console.log(data);
      navigate("/auth/login");
    },
  });
  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const formData = new FormData(e.currentTarget);
    const data = Object.fromEntries(formData) as InputRegister;
    if (data.email && data.password) {
      register({
        variables: { infos: { email: data.email, password: data.password } },
      });
    }
  };
  return (
    <main
      className={`flex min-h-screen flex-col items-center justify-between p-24`}
    >
      <form onSubmit={handleSubmit}>
      <h1 className="mb-5">Inscription</h1>
        <div>
          <input type="text" name="email" placeholder="Indiquez votre email" />
        </div>
        <div>
          <input
            type="password"
            name="password"
            placeholder="Indiquez votre mot de passe"
          />
        </div>
        <input type="submit" />
        <div>
          <span className="text-red-500">{error?.message}</span>
        </div>
      </form>
    </main>
  );
}

export default Register;