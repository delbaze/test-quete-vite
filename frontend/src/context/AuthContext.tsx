import { CHECK_TOKEN } from "@/requetes/queries/auth.queries";
import { CheckTokenInfos, CheckTokenQuery } from "@/types/graphql";
import { useLazyQuery } from "@apollo/client";
import { createContext, PropsWithChildren, useContext, useState } from "react";

type ContextType = {
  infos: {
    email: CheckTokenInfos["email"];
  };
  getInfos(): Promise<void>;
};
const AuthContext = createContext({} as ContextType);

export const useAuth = () => {
  const context = useContext(AuthContext);
  return context;
};

function AuthProvider({ children }: PropsWithChildren) {
  const [state, setState] = useState<{ infos: ContextType["infos"] }>(() => {
    const infos = localStorage.getItem("infos");
    if (infos) {
      return { infos: JSON.parse(infos) };
    }
    return {
      infos: { email: "" },
    };
  });
  const [checkToken] = useLazyQuery<CheckTokenQuery>(CHECK_TOKEN, {
    fetchPolicy: "no-cache",
  });

  const value: ContextType = {
    infos: state.infos,
    getInfos: async () => {
      await checkToken({
        onCompleted(data) {
          const infos = { email: data.checkToken?.email ?? "" };
          console.log("%câ§­", "color: #f200e2", infos);
          setState({ infos });
          localStorage.setItem("infos", JSON.stringify(infos));
        },
      });
    },
  };
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export default AuthProvider;
