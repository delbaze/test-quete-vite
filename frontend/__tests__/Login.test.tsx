import React from "react";
import { describe, it, expect, vi } from "vitest";
import Login from "../src/components/auth/Login";
import * as AuthModule from "../src/context/AuthContext";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { LOGIN } from "../src/requetes/queries/auth.queries";
import { MockedProvider } from "@apollo/client/testing";
import { MemoryRouter } from "react-router-dom";

// Mock de react-router-dom de façon asynchrone
vi.mock("react-router-dom", async (importOriginal) => {
  const actual = (await importOriginal()) as Record<string, unknown>;
  return {
    ...actual,
    useNavigate: () => vi.fn(),
    useLocation: () => ({ state: {} }),
  };
});

// Mock du contexte d'authentification
vi.spyOn(AuthModule, "useAuth").mockReturnValue({
  infos: { email: "" },
  getInfos: vi.fn().mockResolvedValue(undefined),
});

const mocks = [
  {
    request: {
      query: LOGIN,
      variables: {
        infos: { email: "test@mondomaine.com", password: "motdepasse" },
      },
    },
    result: {
      data: {
        login: {
          success: true,
          message: "Bienvenue!",
        },
      },
    },
  },
];

describe("Composant Login", () => {
  it("Snapshot du composant login", async () => {
    const { container, getByText, getByPlaceholderText,  } = render(
      <MemoryRouter initialEntries={["/auth/login"]}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <Login />
        </MockedProvider>
      </MemoryRouter>
    );

    // Remplissage du formulaire avec des données de test
    fireEvent.change(getByPlaceholderText("Indiquez votre email"), {
      target: { value: "test@mondomaine.com" },
    });

    fireEvent.change(getByPlaceholderText("Indiquez votre mot de passe"), {
      target: { value: "motdepasse" },
    });

    // Soumission du formulaire
    fireEvent.submit(getByText("Connexion"));

    // On attend un certain temps pour permettre à onCompleted de s'exécuter
    await waitFor(() => {}, { timeout: 1000 });

    // On vérifie le snapshot après que onCompleted a eu le temps de s'exécuter
    await waitFor(
      () => {
        expect(container).toMatchSnapshot();
      },
      { timeout: 10000 }
    );
  });
}); 