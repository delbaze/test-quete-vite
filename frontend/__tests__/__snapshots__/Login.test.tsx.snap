// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Composant Login > Snapshot du composant login 1`] = `
<div>
  <main
    class="flex min-h-screen flex-col items-center justify-between p-24"
  >
    <form>
      <h1
        class="mb-5"
      >
        Connexion
      </h1>
      <div>
        <input
          name="email"
          placeholder="Indiquez votre email"
          type="text"
        />
      </div>
      <div>
        <input
          name="password"
          placeholder="Indiquez votre mot de passe"
          type="password"
        />
      </div>
      <input
        type="submit"
      />
      <div>
        <span
          class="text-red-500"
        />
        <span
          class="text-blue-500"
        >
          Bienvenue!
        </span>
      </div>
    </form>
  </main>
</div>
`;
